database/dummyData/userEventData.js

```diff
@@ -101,26 +93,131 @@ const generateMovieObj = () => {
+    userId: Math.floor(Math.random() * 10000000),
+    groupId: Math.floor(Math.random() * 2) + 1,
+  };
+  return sessionObj;
+};
+
+const generateFirstQuery = () => {
   const query = {
-    eventId: Math.floor(Math.random() * 4) + 1,
+    eventId: 1,
     movieObj: generateMovieObj(),
     value: Math.random(),
   };
   return query;
 };
 
-const generateEvents = (num) => {
-  const eventArray = [];
-  for (let i = 0; i < num; i += 1) {
-    const event = {
+const generateEvent = (prevEvent = null) => {
+  if (!prevEvent) {
+    // User is logging in to a new session
+    return {
       session: generateSession(),
-      query: generateQuery(),
+      query: generateFirstQuery(),
     };
-    eventArray.push(event);
   }
-  eventArray.map(event => db.addEvent(event.session, event.query));
-  return Promise.all(eventArray);
+  const { eventId } = prevEvent.query;
+  let newEventId = null;
+  let newMovieObj = null;
+  if (eventId === 1) {
+    // User has already logged in
+    // 95% chance of watching a movie, 5% of logging out
+    newEventId = Math.random() < 0.05 ? 4 : 2;
+  } else if (eventId === 2) {
+    // User is watching a movie
+    // 80% chance of stopping movie, 20% of logging out directly
+    newEventId = Math.random() < 0.8 ? 3 : 4;
+  } else {
+    // User has stopped watching a movie
+    // 50% chance of starting a new movie, 50% of logging out
+    newEventId = Math.random() < 0.5 ? 2 : 4;
+    if (newEventId === 2) {
+      newMovieObj = generateMovieObj();
+    }
+  }
+  return {
+    session: prevEvent.session,
+    query: {
+      eventId: newEventId,
+      movieObj: newMovieObj === null ? prevEvent.query.movieObj : newMovieObj,
+      isRec: newMovieObj === null ? prevEvent.query.isRec : Math.round(Math.random()) === 0,
+      value: Math.random(), // Refine and refactor
+    },
+  };
+};
+
+const generateUserSession = () => {
+  const userSession = [];
+  let event = generateEvent();
+  userSession.push(event);
+  while (event.query.eventId !== 4) {
+    event = generateEvent(event);
+    userSession.push(event);
+  }
+  return userSession;
+};
+
+const generateAllSessions = (num) => {
+  const totalSessions = [];
+  for (let i = 0; i < num; i += 1) {
+    totalSessions.push(generateUserSession());
+  }
+  return totalSessions;
+};
+
+const simulateUserEvents = (numOfSessions) => {
+  const totalSessions = generateAllSessions(numOfSessions);
+  // holds all of the events, in chronological order, that will be added to the database
+  const eventArray = [];
+  while (totalSessions.length > 0) {
+    const randIndex = Math.floor(Math.random() * totalSessions.length);
+    eventArray.push(totalSessions[randIndex].shift());
+    if (totalSessions[randIndex].length === 0) {
+      totalSessions.splice(randIndex, 1);
+    }
+  }
+  return eventArray;
+};
+
+const eventArray = simulateUserEvents(20);
+const eventDashboard = [];
+
+const generateEvents = (num = 1) => {
+  const event = eventArray[num - 1];
```

Comment:
Is there any reason for creating eventArray?  Can you just replace eventArray[num - 1] with simulateUserEvents(num)?

Reply:
eventArray, created on line 185 needs to be initialized outside of the generateEvents function so that it is not redefined in that recursive function each time it runs.